### PR TITLE
Feature/font carrier

### PR DESCRIPTION
**Pull request by zFitness**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## font-carrier ([<kbd>unpkg</kbd>](https://unpkg.com/browse/font-carrier@latest/))
was missing the following properties:
1. engine
1. helper
1. easySvg
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against ebbe2aaa949ed567d9ec6522b897cfe72718f602
</p>

File: types/font-carrier/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "rules": {
+        "dt-header": "off"
```
Comment (weswigham) [inline types/font-carrier/.eslintrc.json:3]: Please don't disable this or any other lint rule, for the header specifically, we use metadata from the header for automation, thanks~

**Review by weswigham — CHANGES_REQUESTED** [review]

File: types/font-carrier/tslint.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+    "extends": "@definitelytyped/dtslint/dt.json",
+    "rules": {
+        "array-type": false,
```
Comment (weswigham) [inline types/font-carrier/tslint.json:4]: Please don't disable this or any other lint rule for new definitions, thanks~

File: types/font-carrier/.eslintrc.json (hunk 1 above)
Comment (zFitness) [reply types/font-carrier/.eslintrc.json]: ok

File: types/font-carrier/tslint.json (hunk 2 above)
Comment (zFitness) [reply types/font-carrier/tslint.json]: hello， I have dealt with the problem.

**Review by zFitness — COMMENTED** [review]
don't disable this or any other lint rule for new definitions

File: types/font-carrier/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
+// Type definitions for font-carrier 0.3
+// Project: https://github.com/purplebamboo/font-carrier#readme
+// Definitions by: zfitness <https://github.com/zFitness>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+// tslint:disable-next-line: no-declare-current-package no-single-declare-module
```
Comment (weswigham) [inline types/font-carrier/index.d.ts:8]: You should just be able to delete this `declare module` wrapper and this lint exception.

Comment (zFitness) [reply types/font-carrier/index.d.ts]: ok

Comment (zFitness) [reply types/font-carrier/index.d.ts]: Now , i dont want to export Base<T>, but  it reported an error： class Base<T>
Top-level declarations in .d.ts files must start with either a 'declare' or 'export' modifier

Comment (weswigham) [reply types/font-carrier/index.d.ts]: Just add the `declare`.

Comment (zFitness) [reply types/font-carrier/index.d.ts]: It's a small problem， please merge

**Review by zFitness — COMMENTED** [review]
declare

File: types/font-carrier/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+export type FontType = 'ttf' | 'eot' | 'woff' | 'woff2' | 'svg';
+
+export interface FileOptions {
+    /** 支持文件地址，也支持文件 buffer */
+    path?: string | Buffer;
+    /** 导出的字体类型 */
+    types?: FontType[];
+}
+
+export class Base<T> {
```
Comment (peterblazejewicz) [inline types/font-carrier/index.d.ts:17]: correct me, if I'm wrong, but the JS module does not export any of such symbols:
https://github.com/purplebamboo/font-carrier/blob/master/lib/index.js
one should avoid creating symbols that does not exists in original. Common interface can be achiveve via, say, interface. Or just keep this private within DT module using `export {};`.

Comment (zFitness) [reply types/font-carrier/index.d.ts]: hello, i add a base.d.ts

Comment (zFitness) [reply types/font-carrier/index.d.ts]: please

**Review by zFitness — COMMENTED** [review]
1

File: types/font-carrier/class/font.d.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+export interface FileOptions {
+    /** 支持文件地址，也支持文件 buffer */
+    path?: string | Buffer;
+    /** 导出的字体类型 */
+    types?: FontType[];
+}
+
+/** 一个 font 对象就表示解析后的字体对象，包含字体所有的信息 */
+export class Font extends Base<FontOptions> {
```
Comment (sheetalkamat) [inline types/font-carrier/class/font.d.ts:23]: Does not export class Font directly but through `export = ` per https://unpkg.com/browse/font-carrier@0.3.1/lib/class/font.js

You need to have `declare class Font` and then `export = Font` instead.

Comment (sheetalkamat) [reply types/font-carrier/class/font.d.ts:23]: Same goes for all the other files in `front-carrier/class`

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Comment by peterblazejewicz** [comment]
@zFitness to wrap up: 
- this seems to be regular CommonJS module, using CJS exports/imports, not ES6 import/exports
- in declaration here, stick to the CommonJS module syntax:
https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html#common-commonjs-patterns

File: types/font-carrier/class/font.d.ts (hunk 5 above)
Comment (zFitness) [reply types/font-carrier/class/font.d.ts]: thanks, Please check it again

File: types/font-carrier/class/glyph.d.ts [hunk 6]
```diff
@@ -39,3 +40,6 @@ export class Glyph extends Base<GlyphOptions> {
         }>,
     ): string;
 }
+
+export = Glyph;
+export as namespace Glyph;
```
Comment (peterblazejewicz) [inline types/font-carrier/class/glyph.d.ts]:
```suggestion
```
IMO this is CommonJS module, without support for global script in distributed package, like wth UMD modules (no code loaded via `script` tag)

File: types/font-carrier/class/fontface.d.ts [hunk 7]
```diff
@@ -1,19 +1,23 @@
...
+        descent: string;
+    }
 }
-
 /** fontface 对象包含了字体的一些特殊相关信息 */
-export class FontFace extends Base<FontFaceOptions> {}
+declare class FontFace extends Base<FontFace.FontFaceOptions> {}
+
+export = FontFace;
+export as namespace FontFace;
```
Comment (peterblazejewicz) [inline types/font-carrier/class/fontface.d.ts]:
```suggestion
```
ditto about global script

File: types/font-carrier/class/font.d.ts [hunk 8]
```diff
@@ -52,5 +54,8 @@ export class Font extends Base<FontOptions> {
      * 导出字体
      * @param options
      */
-    output(options?: FileOptions): Buffer[];
+    output(options?: Font.FileOptions): Buffer[];
 }
+
+export = Font;
+export as namespace Font;
```
Comment (peterblazejewicz) [inline types/font-carrier/class/font.d.ts]:
```suggestion
```
ditto about global script

**Review by zFitness — COMMENTED** [review]
1

File: types/font-carrier/class/font.d.ts (hunk 8 above)
Comment (zFitness) [reply types/font-carrier/class/font.d.ts]: hello,  i have delete it, and commit yesterday

**Review by sheetalkamat — APPROVED** [review]

File: types/font-carrier/class/glyph.d.ts (hunk 6 above)
Comment (peterblazejewicz) [inline types/font-carrier/class/glyph.d.ts]:
```suggestion
```

**Review by peterblazejewicz — APPROVED** [review]

**Review by sheetalkamat — APPROVED** [review]

**Comment by zFitness** [comment]
Ready to merge

**Pull request merged**